internal/dag/cache_test.go

```diff
@@ -2751,3 +2920,235 @@ func TestLookupUpstreamValidation(t *testing.T) {
+				TargetRef: targetRef("", "Service", serviceName, namespace, sectionName),
+				TLS: gatewayapi_v1alpha2.BackendTLSPolicyConfig{
+					CACertRefs: []gatewayapi_v1beta1.LocalObjectReference{
+						{
+							Group: "",
+							Kind:  "Secret",
+							Name:  "ca",
+						},
+					},
+					Hostname: "example.com",
+				},
+			},
+		}
+	}
+
+	tests := map[string]struct {
+		targetRef          gatewayapi_v1alpha2.PolicyTargetReferenceWithSectionName
+		backendTLSPolicies []*gatewayapi_v1alpha2.BackendTLSPolicy
+		want               *gatewayapi_v1alpha2.BackendTLSPolicy
+		wantFound          bool
+	}{
+		"finds the BackendTLSPolicy with the matching targetRef": {
+			targetRef: targetRef("", "Service", "backend-service", nil, nil),
+			backendTLSPolicies: []*gatewayapi_v1alpha2.BackendTLSPolicy{
+				backendTLSPolicy("btp", "backend-service", nil, nil),
+				backendTLSPolicy("btp1", "backend-service-with-section-name", nil, ptr.To("https")),
+				backendTLSPolicy("btp2", "backend-service-with-section-name", nil, ptr.To("https2")),
+			},
+			want:      backendTLSPolicy("btp", "backend-service", nil, nil),
+			wantFound: true,
+		},
+		"finds the BackendTLSPolicy matching targetRef with section name": {
+			targetRef: targetRef("", "Service", "backend-service-with-section-name", nil, ptr.To("https2")),
+			backendTLSPolicies: []*gatewayapi_v1alpha2.BackendTLSPolicy{
+				backendTLSPolicy("btp", "backend-service", nil, nil),
+				backendTLSPolicy("btp1", "backend-service-with-section-name", nil, ptr.To("https")),
+				backendTLSPolicy("btp2", "backend-service-with-section-name", nil, ptr.To("https2")),
+			},
+			want:      backendTLSPolicy("btp2", "backend-service-with-section-name", nil, ptr.To("https2")),
+			wantFound: true,
+		},
+		"finds the fallback BackendTLSPolicy matching targetRef but not section name": {
+			targetRef: targetRef("", "Service", "backend-service-with-fallback", nil, ptr.To("https2")),
+			backendTLSPolicies: []*gatewayapi_v1alpha2.BackendTLSPolicy{
+				backendTLSPolicy("btp", "backend-service", nil, nil),
+				backendTLSPolicy("btp1", "backend-service-with-fallback", nil, nil),
+				backendTLSPolicy("btp2", "backend-service-with-fallback", nil, ptr.To("https")),
+			},
+			want:      backendTLSPolicy("btp1", "backend-service-with-fallback", nil, nil),
+			wantFound: true,
+		},
+		"finds the fallback BackendTLSPolicy matching targetRef with section name": {
+			targetRef: targetRef("", "Service", "backend-service-with-fallback", nil, ptr.To("https")),
+			backendTLSPolicies: []*gatewayapi_v1alpha2.BackendTLSPolicy{
+				backendTLSPolicy("btp", "backend-service", nil, nil),
+				backendTLSPolicy("btp1", "backend-service-with-fallback", nil, nil),
+				backendTLSPolicy("btp2", "backend-service-with-fallback", nil, ptr.To("https")),
+			},
+			want:      backendTLSPolicy("btp2", "backend-service-with-fallback", nil, ptr.To("https")),
+			wantFound: true,
+		},
+		"finds the BackendTLSPolicy matching namespace": {
```

Comment:
i think there should be some consideration of what namespace the backendtlspolicy resides in as well, at least as an initial pass let's not allow policies that live in a different namespace from the Service (for the purposes of this test, the targetRef target namespace) they reference

so we may need to always construct a targetRef with a namespace which in the processor this would be the namespace of the HTTPRoute if not specified on the backendref, otherwise the ns of the backendref

Reply:
Yeah logic here has some hidden complexity. I did a bit of a rewrite, now matching the namespace provided on the targetref to the backendtlspolicy namespace instead, then check that the backendtlspolicy namespace matches the spec.targetRef.namespace afterward. Also added some logic around default namespaces and empty namespaces e.g if spec.targetRef.namespace is empty that should be fine since that should be used for cross namespace-ing, which is unsupported. Added additional testing here as well, but could always use another look to check if anything is missed or unclear.